src/libraries/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/StreamMemoryBlockProvider.cs

```diff
@@ -76,9 +76,11 @@ internal static unsafe NativeHeapMemoryBlock ReadMemoryBlockNoLock(Stream stream
             {
                 stream.Seek(start, SeekOrigin.Begin);
 
-                if (!isFileStream || !FileStreamReadLightUp.TryReadFile(stream, block.Pointer, start, size))
+                int bytesRead = 0;
+
+                if (!isFileStream || !FileStreamReadLightUp.TryReadFile(stream, block.Pointer, size, out bytesRead))
                 {
-                    stream.CopyTo(block.Pointer, size);
+                    stream.CopyTo(block.Pointer + bytesRead, size - bytesRead);
```

Comment:
I see this logic looks correct but I find it very confusing that you progress pointer when TryReadFile returns false. It seems to me TryReadFile should return false only when nothing is read or if error occurred... I suspect that's why original error happened in the first place. I'd suggest to not progress and instead return false only when no bytes are read

Reply:
> I see this logic looks correct but I find it very confusing that you progress pointer when TryReadFile returns false. 

I agree. The problem is that if `FileStream.Read` succeeds only partially, it returns fewer bytes than requested and also moves the file position. So the next read, will start from the updated position. Example:

```cs
FileStream fs = File.Open(path);
byte[] bytes = new byte[2];
int bytesRead = fs.Read(bytes.AsSpan());
// now we have 3 possibilitites:

if (bytesRead == 2)
{
    // success, we have read as many bytes as we wanted
}
else if (bytesRead == 1)
{
   // partial success, we have read something, but less than requested
   // the FileStream.Position is now set to 1
   // and the next read will start from this position
}
else if (bytesRead == 0)
{
   // obvious failure
}
```

I could handle it by calling  `FileStream.Seek(-bytesRead, SeekOrigin.Current)` on incomplete read, but this would be one extra method call (in the past it would be also an additional syscall)

@krwq which option do you prefer?

Reply:
Perhaps just rename to something like TryFullyRead, that would be enough to at least someone to see what they mean by `Fully`